cartography/intel/okta/oktaintel.py

```diff
@@ -0,0 +1,1193 @@
+# Okta intel module
+import json
+import logging
+import os
+
+from okta import AppInstanceClient
+from okta import FactorsClient
+from okta import UserGroupsClient
+from okta import UsersClient
+from okta.framework.ApiClient import ApiClient
+from okta.framework.OktaError import OktaError
+from okta.framework.PagedResults import PagedResults
+from okta.models.usergroup import UserGroup
+
+from cartography.config import Config
+from cartography.util import run_cleanup_job
+
+logger = logging.getLogger(__name__)
+
+OKTA_API_KEY = os.environ.get('CREDENTIALS_OKTA_API_KEY')
+
+OKTA_ORG_ID = os.environ.get("OKTA_ORG_ID")
+
+
+def _create_user_client(okta_org):
+    """
+    Create Okta User Client
+    :param okta_org: Okta organization name
+    :return: Instance of UsersClient
+    """
+    # https://github.com/okta/okta-sdk-python/blob/master/okta/models/user/User.py
+    user_client = UsersClient(
+        base_url=f"https://{okta_org}.okta.com/",
+        api_token=OKTA_API_KEY,
+    )
+
+    return user_client
+
+
+def _create_group_client(okta_org):
+    """
+    Create Okta UserGroupsClient
+    :param okta_org: Okta organization name
+    :return: Instance of UserGroupsClient
+    """
+    usergroups_client = UserGroupsClient(
+        base_url=f"https://{okta_org}.okta.com/",
+        api_token=OKTA_API_KEY,
+    )
+
+    return usergroups_client
+
+
+def _create_application_client(okta_org):
+    """
+    Create Okta AppInstanceClient
+    :param okta_org: Okta organization name
+    :return: Instance of AppInstanceClient
+    """
+    app_client = AppInstanceClient(
+        base_url=f"https://{okta_org}.okta.com/",
+        api_token=OKTA_API_KEY,
+    )
+
+    return app_client
+
+
+def _create_factor_client(okta_org):
+    """
+    Create Okta FactorsClient
+    :param okta_org: Okta organization name
+    :return: Instance of FactorsClient
+    """
+
+    # https://github.com/okta/okta-sdk-python/blob/master/okta/FactorsClient.py
+    factor_client = FactorsClient(
+        base_url=f"https://{okta_org}.okta.com/",
+        api_token=OKTA_API_KEY,
+    )
+
+    return factor_client
+
+
+def _create_api_client(okta_org, path_name):
+    """
+    Create Okta ApiClient
+    :param okta_org: Okta organization name
+    :param path_name: API Path
+    :return: Instance of ApiClient
+    """
+    api_client = ApiClient(
+        base_url=f"https://{okta_org}.okta.com/",
+        pathname=path_name,
+        api_token=OKTA_API_KEY,
+    )
+
+    return api_client
+
+
+def _get_okta_users(user_client):
+    """
+    Get Okta users from Okta server
+    :param user_client: user client
+    :return: Array of dictionary containing user properties
+    """
+    user_list = []
+    paged_users = user_client.get_paged_users()
+    while True:
+        for current_user in paged_users.result:
+            user_props = transform_okta_user(current_user)
```

Comment:
Please refactor this function such that it does not call transform directly, rather it should return the raw results from the API which can then be passed to the transform function by the caller. This is the pattern that is used through the rest of the codebase and it was selected to make the code easier to test and debug.

Reply:
This applies to all the _get functions below as well.